cbmc/proofs/proof_guide.md

```diff
@@ -0,0 +1,399 @@
+d[//]: # (SPDX-License-Identifier: CC-BY-4.0)
+
+# CBMC Proof Guide and Cookbook for MLKEM-C
+
+This doc acts as a guide to developing proofs of C code using CBMC. It concentrates
+on the use of _contracts_ to achieve _unbounded_ and _modular_ proofs of type-safety
+and correctness properties.
+
+This document uses the abbreviated forms of the CBMC contracts defined by macros in the
+[cbmc.h](../../mlkem/cbmc.h) header file in the MLKEM-C sources.
+
+## Common Proof Patterns
+
+### Loops (general advice)
+
+1. A function should contain at most one outermost loop statement. If the function you're trying to prove has more than one outermost loop, then re-factor it into two or more functions.
+
+2. The one outermost loop statement should be _final_ statement before the function returns. Don't have complicated code _after_ the loop body.
+
+### For loops
+
+The most common, and easiest, patten is a "for" loop that has a counter starting at 0, and counting up to some upper bound, like this:
+
+```
+for (int i = 0; i < C; i++) {
+    S;
+}
+```
+Notes:
+1. It is good practice to declare the counter variable locally, within the scope of the loop.
+2. The counter variable should be a constant within the loop body. In the example above, do _not_ modify `i` in the body of the loop.
+3. "int" is the best type for the counter variable. "unsigned" integer types complicate matters with their modulo N semantics. Avoid "size_t" since its large range (possibly unsigned 64 bit) can slow proofs down.
+
+CBMC requires basic assigns, loop-invariant, and decreases contracts _in exactly that order_. Note that the contracts appear _before_ the opening `{` of the loop body, so we also need to tell `clang-format` not to reformat these lines. The basic pattern is thus:
+
+```
+for (int i = 0; i < C; i++)
+    // clang-format off
+ASSIGNS(i) // plus whatever else S does
+INVARIANT(i >= 0)
+INVARIANT(i <= C)
+DECREASES(C - i)
+// clang-format on
+{
+    S;
+}
+```
+
+The `i <= C` in the invariant is _not_ a typo. CBMC places the invariant just _after_ the loop counter has been incremented, but just _before_ the loop exit test, so it is possible for `i == C` at the invariant on the final iteration of the loop.
+
+### Iterating over an array for a for loop
+
+A common pattern - doing something to every element of an array. An example would be setting every element of a byte-array to 0x00 given a pointer to the first element and a length. Initially, we want to prove type safety of this function, so we won't even bother with a post-condition. The function specification might look like this:
+
+```
+void zero_array_ts (uint8_t *dst, int len)
+    // clang-format off
+REQUIRES(IS_FRESH(dst, len))
+ASSIGNS(OBJECT_WHOLE(dst));
```

Comment:
Could you add some words about what `IS_FRESH` and `OBJECT_WHOLE` are doing, please?

Reply:
Added words to that effect.